File: backend/services/network.js
export const NETWORK_PATH_LIMIT = 1000;

// Include the root even when it has no relationships. Search both directions;
// ordering shortest paths first keeps hop distances correct when capped.
// Crime types describe cases but must not bridge otherwise unrelated cases.
// Exclude them from every position except the final node in a path.
export const NETWORK_QUERY = `
  MATCH path = (root:Person {person_id: $id})-[*0..4]-(connected)
  WHERE none(node IN nodes(path)[0..-1] WHERE node:CrimeType)
  RETURN path
  ORDER BY length(path), elementId(connected)
  LIMIT ${NETWORK_PATH_LIMIT + 1}
`;

export async function fetchNetwork(id, runCypher) {
  const records = await runCypher(NETWORK_QUERY, { id });
  if (records.length === 0) return null;
  return {
    ...serializeNetwork(records.slice(0, NETWORK_PATH_LIMIT)),
    truncated: records.length > NETWORK_PATH_LIMIT,
    pathLimit: NETWORK_PATH_LIMIT,
  };
}

// Preserve Neo4j element IDs so parallel relationships and node types stay distinct.
export function serializeNetwork(records) {
  const nodes = new Map();
  const edges = new Map();
  for (const { path } of records) {
    const addNode = (node, depth) => {
      const existing = nodes.get(node.elementId);
      if (existing) { existing.depth = Math.min(existing.depth, depth); return; }
      const p = node.properties;
      const name = p.name || p.city || p.crime_name || p.registration || p.case_id || p.person_id || node.labels[0];
      nodes.set(node.elementId, {
        id: node.elementId, kind: node.labels[0] || 'Record', depth,
        label: name, displayLabel: p.name ? `${name}\n${p.alias ? `“${p.alias}”` : p.person_id || ''}` : name,
        personId: p.person_id == null ? null : String(p.person_id), alias: p.alias || null, city: p.city || null,
        provenance: p.source || null,
      });
    };
    addNode(path.start, 0);
    path.segments.forEach((segment, index) => {
      addNode(segment.end, index + 1);
      const r = segment.relationship;
      edges.set(r.elementId, {
        id: r.elementId, source: r.startNodeElementId, target: r.endNodeElementId,
        label: r.type.replaceAll('_', ' '), provenance: r.properties.source || null,
        reason: r.properties.reason || null,
      });
    });
  }
  return { nodes: [...nodes.values()], edges: [...edges.values()] };
}

File: backend/services/network.py
NETWORK_PATH_LIMIT = 1000
NETWORK_QUERY = f"""
  MATCH path = (root:Person {{person_id: $id}})-[*0..4]-(connected)
  WHERE none(node IN nodes(path)[0..-1] WHERE node:CrimeType)
  RETURN path
  ORDER BY length(path), elementId(connected)
  LIMIT {NETWORK_PATH_LIMIT + 1}
"""


async def fetch_network(person_id, run_cypher):
    records = await run_cypher(NETWORK_QUERY, {"id": person_id})
    if not records:
        return None
    return {
        **serialize_network(records[:NETWORK_PATH_LIMIT]),
        "truncated": len(records) > NETWORK_PATH_LIMIT,
        "pathLimit": NETWORK_PATH_LIMIT,
    }


def serialize_network(records):
    nodes, edges = {}, {}
    for record in records:
        path = record["path"]
        # Python's Path.nodes follows traversal order, even for incoming edges.
        for depth, node in enumerate(path.nodes):
            if node.element_id in nodes:
                nodes[node.element_id]["depth"] = min(nodes[node.element_id]["depth"], depth)
                continue
            kind = next(
                (
                    label
                    for label in ("Person", "Case", "Location", "CrimeType", "Vehicle")
                    if label in node.labels
                ),
                next(iter(sorted(node.labels)), "Record"),
            )
            name = next(
                (
                    node.get(key)
                    for key in (
                        "name",
                        "city",
                        "crime_name",
                        "registration",
                        "case_id",
                        "person_id",
                    )
                    if node.get(key)
                ),
                kind,
            )
            subtitle = f"“{node['alias']}”" if node.get("alias") else node.get("person_id") or ""
            nodes[node.element_id] = {
                "id": node.element_id,
                "kind": kind,
                "depth": depth,
                "label": name,
                "displayLabel": f"{name}\n{subtitle}" if node.get("name") else name,
                "personId": str(node["person_id"]) if node.get("person_id") is not None else None,
                "alias": node.get("alias") or None,
                "city": node.get("city") or None,
                "provenance": node.get("source") or None,
            }
        for relationship in path.relationships:
            edges[relationship.element_id] = {
                "id": relationship.element_id,
                "source": relationship.start_node.element_id,
                "target": relationship.end_node.element_id,
                "label": relationship.type.replace("_", " "),
                "provenance": relationship.get("source") or None,
                "reason": relationship.get("reason") or None,
            }
    return {"nodes": list(nodes.values()), "edges": list(edges.values())}
